fs-cache/src/cache.rs

```diff
@@ -0,0 +1,668 @@
+                    format!("Failed to write value for key {}: {}", key, err),
+                )
+            })
+    }
+
+    /// Reads a value from disk.
+    fn read_from_disk(&self, key: &K) -> Result<V>
+    where
+        V: From<Vec<u8>>, // Add trait bound for reading binary data
+    {
+        let file_path = self.path.join(key.to_string());
+        let contents = fs::read(&file_path)?;
+        Ok(V::from(contents))
+    }
+
+    /// Returns the size of a value in bytes.
+    ///
+    /// First checks the memory cache for size information to avoid disk access.
+    /// Falls back to checking the file size on disk if not found in memory.
+    fn get_file_size(&self, key: &K) -> Result<usize> {
+        Ok(fs::metadata(self.path.join(key.to_string()))?.len() as usize)
+    }
+
+    /// Adds or updates a value in the memory cache, evicting old entries if needed.
+    /// Returns error if value is larger than maximum memory limit.
+    fn update_memory_cache(&mut self, key: &K, value: &V) -> Result<()> {
+        log::debug!("cache/{}: caching in memory for key {}", self.label, key);
+        let size = self.get_file_size(key)?;
+
+        // If single value is larger than total limit, just skip memory caching
+        if size > self.max_memory_bytes {
+            log::error!(
+                "cache/{}: value size {} exceeds limit {}",
+                self.label,
+                size,
+                self.max_memory_bytes
+            );
+            return Ok(());
+        }
+
+        // Remove oldest entries until we have space for new value
+        while self.current_memory_bytes + size > self.max_memory_bytes {
+            let (_, old_entry) = self
+                .memory_cache
+                .pop_lru()
+                .expect("Cache should have entries to evict");
+            debug_assert!(
+                self.current_memory_bytes >= old_entry.size,
+                "Memory tracking inconsistency detected"
+            );
+            self.current_memory_bytes = self
+                .current_memory_bytes
+                .saturating_sub(old_entry.size);
+        }
+
+        // Add new value and update size
+        self.memory_cache.put(
+            key.clone(),
+            CacheEntry {
+                value: value.clone(),
+                size,
+            },
+        );
+        self.current_memory_bytes += size;
+
+        log::debug!(
+            "cache/{}: added {} bytes, total {}/{}",
+            self.label,
+            size,
+            self.current_memory_bytes,
+            self.max_memory_bytes
+        );
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rstest::{fixture, rstest};
+    use std::{
+        fs::File,
+        io::Write,
+        time::{Duration, SystemTime},
+    };
+    use tempdir::TempDir;
+
+    // Helper struct that implements required traits
+    #[derive(Clone, Debug, PartialEq)]
+    struct TestValue(Vec<u8>);
```

Comment:
I’m fairly confident we can use `Vec<u8>` directly instead of creating a new struct like `TestValue`. I tried it locally and it works, so I think it’s better to go with `Vec<u8>` directly.